File: lib/grex/builder.ex
defmodule Grex.Builder do
  defstruct [:inner]

  def build(builder) do
    Grex.Native.builder_build(builder)
  end

  def build_and_compile(builder) do
    builder
    |> build()
    |> Regex.compile()
  end

  def build_and_compile!(builder) do
    builder
    |> build()
    |> Regex.compile!()
  end

  def with_capturing_groups(builder) do
    Grex.Native.builder_with_capturing_groups(builder)
  end

  def with_case_insensitive_matching(builder) do
    Grex.Native.builder_with_case_insensitive_matching(builder)
  end

  def with_conversion_of_digits(builder) do
    Grex.Native.builder_with_conversion_of_digits(builder)
  end

  def with_conversion_of_non_digits(builder) do
    Grex.Native.builder_with_conversion_of_non_digits(builder)
  end

  def with_conversion_of_non_whitespace(builder) do
    Grex.Native.builder_with_conversion_of_non_whitespace(builder)
  end

  def with_conversion_of_non_words(builder) do
    Grex.Native.builder_with_conversion_of_non_words(builder)
  end

  def with_conversion_of_repetitions(builder) do
    Grex.Native.builder_with_conversion_of_repetitions(builder)
  end

  def with_conversion_of_whitespace(builder) do
    Grex.Native.builder_with_conversion_of_whitespace(builder)
  end

  def with_conversion_of_words(builder) do
    Grex.Native.builder_with_conversion_of_words(builder)
  end

  def with_escaping_of_non_ascii_chars(builder, use_surrogate_pairs) do
    Grex.Native.builder_with_escaping_of_non_ascii_chars(builder, use_surrogate_pairs)
  end

  def with_minimum_repetitions(builder, quantity) do
    Grex.Native.builder_with_minimum_repetitions(builder, quantity)
  end

  def with_minimum_substring_length(builder, length) do
    Grex.Native.builder_with_minimum_substring_length(builder, length)
  end

  def with_verbose_mode(builder) do
    Grex.Native.builder_with_verbose_mode(builder)
  end

  def without_anchors(builder) do
    Grex.Native.builder_without_anchors(builder)
  end

  def without_end_anchor(builder) do
    Grex.Native.builder_without_end_anchor(builder)
  end

  def without_start_anchor(builder) do
    Grex.Native.builder_without_start_anchor(builder)
  end
end

File: native/grex_native/src/builder.rs
use grex::RegExpBuilder;
use rustler::{NifStruct, ResourceArc};
use std::sync::RwLock;

pub struct RegExpBuilderWrapper(RwLock<RegExpBuilder>);

#[derive(NifStruct)]
#[module = "Grex.Builder"]
pub struct Builder {
    pub inner: ResourceArc<RegExpBuilderWrapper>,
}

#[rustler::nif]
pub fn builder_new(input: Vec<String>) -> rustler::NifResult<Builder> {
    if input.is_empty() {
        return Err(rustler::Error::BadArg);
    }

    Ok(Builder {
        inner: ResourceArc::new(RegExpBuilderWrapper(RwLock::new(RegExpBuilder::from(
            &input,
        )))),
    })
}

#[rustler::nif]
pub fn builder_build(input: Builder) -> String {
    let mut builder = input.inner.0.write().unwrap();
    builder.build()
}

#[rustler::nif]
pub fn builder_with_capturing_groups(input: Builder) -> Builder {
    {
        let mut builder = input.inner.0.write().unwrap();
        builder.with_capturing_groups();
    }

    input
}

#[rustler::nif]
pub fn builder_with_case_insensitive_matching(input: Builder) -> Builder {
    {
        let mut builder = input.inner.0.write().unwrap();
        builder.with_case_insensitive_matching();
    }

    input
}

#[rustler::nif]
pub fn builder_with_conversion_of_digits(input: Builder) -> Builder {
    {
        let mut builder = input.inner.0.write().unwrap();
        builder.with_conversion_of_digits();
    }

    input
}

#[rustler::nif]
pub fn builder_with_conversion_of_non_digits(input: Builder) -> Builder {
    {
        let mut builder = input.inner.0.write().unwrap();
        builder.with_conversion_of_non_digits();
    }

    input
}

#[rustler::nif]
pub fn builder_with_conversion_of_non_whitespace(input: Builder) -> Builder {
    {
        let mut builder = input.inner.0.write().unwrap();
        builder.with_conversion_of_non_whitespace();
    }

    input
}

#[rustler::nif]
pub fn builder_with_conversion_of_non_words(input: Builder) -> Builder {
    {
        let mut builder = input.inner.0.write().unwrap();
        builder.with_conversion_of_non_words();
    }

    input
}

#[rustler::nif]
pub fn builder_with_conversion_of_repetitions(input: Builder) -> Builder {
    {
        let mut builder = input.inner.0.write().unwrap();
        builder.with_conversion_of_repetitions();
    }

    input
}

#[rustler::nif]
pub fn builder_with_conversion_of_whitespace(input: Builder) -> Builder {
    {
        let mut builder = input.inner.0.write().unwrap();
        builder.with_conversion_of_whitespace();
    }

    input
}

#[rustler::nif]
pub fn builder_with_conversion_of_words(input: Builder) -> Builder {
    {
        let mut builder = input.inner.0.write().unwrap();
        builder.with_conversion_of_words();
    }

    input
}

#[rustler::nif]
pub fn builder_with_escaping_of_non_ascii_chars(
    input: Builder,
    use_surrogate_pairs: bool,
) -> Builder {
    {
        let mut builder = input.inner.0.write().unwrap();
        builder.with_escaping_of_non_ascii_chars(use_surrogate_pairs);
    }

    input
}

#[rustler::nif]
pub fn builder_with_minimum_repetitions(
    input: Builder,
    quantity: u32,
) -> rustler::NifResult<Builder> {
    if quantity == 0 {
        return Err(rustler::Error::BadArg);
    }

    {
        let mut builder = input.inner.0.write().unwrap();
        builder.with_minimum_repetitions(quantity);
    }

    Ok(input)
}

#[rustler::nif]
pub fn builder_with_minimum_substring_length(
    input: Builder,
    length: u32,
) -> rustler::NifResult<Builder> {
    if length == 0 {
        return Err(rustler::Error::BadArg);
    }

    {
        let mut builder = input.inner.0.write().unwrap();
        builder.with_minimum_substring_length(length);
    }

    Ok(input)
}

#[rustler::nif]
pub fn builder_with_verbose_mode(input: Builder) -> Builder {
    {
        let mut builder = input.inner.0.write().unwrap();
        builder.with_verbose_mode();
    }

    input
}

#[rustler::nif]
pub fn builder_without_anchors(input: Builder) -> Builder {
    {
        let mut builder = input.inner.0.write().unwrap();
        builder.without_anchors();
    }

    input
}

#[rustler::nif]
pub fn builder_without_end_anchor(input: Builder) -> Builder {
    {
        let mut builder = input.inner.0.write().unwrap();
        builder.without_end_anchor();
    }

    input
}

#[rustler::nif]
pub fn builder_without_start_anchor(input: Builder) -> Builder {
    {
        let mut builder = input.inner.0.write().unwrap();
        builder.without_start_anchor();
    }

    input
}

pub fn load(env: rustler::Env, _: rustler::Term) -> bool {
    rustler::resource!(RegExpBuilderWrapper, env);
    true
}
